### ISSUE TITLE
FLIP: rm 'shape' from Module.param call signature 

### ISSUE THREAD

**Issue by j-towns** [google/flax#160]
### Goals

In our current implementation, `initializer` methods (which get passed to [`Module.param`](https://github.com/google/flax/blob/8bf6e36d5209d74296bfaeeb6a505ce226b19b8a/flax/nn/base.py#L514)) are required to have the following signature:
```python
def initializer(rng_key, shape):
  # ...
  return initialized_parameters
```
and in `Module.param` we assert that `initialized_parameters.shape == shape`.

Sometimes an initializer needs more (or less) information than the shape of its output, and at the moment this is achieved by writing the initializer function within a `Module` definition so that it can close over other data that it requires. For example, weightnorm initialization, which is data-dependent, can be implemented as follows, note the necessity of the dummy shape arguments:

```python
class Conv2D(nn.Module):
  def apply(self, inputs, features, kernel_size):
    strides = (inputs.ndim - 2) * (1,)

    conv = partial(
        lax.conv_general_dilated, window_strides=strides, padding='VALID',
        dimension_numbers=('NHWC', 'HWIO', 'NHWC'))

    in_features = inputs.shape[-1]
    kernel_shape = kernel_size + (in_features, features)

    def initializer(key, shape):
      # A weightnorm initializer generating a (direction, scale, bias) tuple.
      # Note that the shape argument is not used.
      direction = nn.initializers.normal()(key, kernel_shape)
      unnormed_out = conv(inputs, _l2_normalize(direction))
      mean = np.mean(unnormed_out, (0, 1, 2))
      var  = np.std (unnormed_out, (0, 1, 2))
      return dict(direction=direction, scale=1 / var, bias=-mean / var)

    # We feed in None as a dummy shape argument to self.param.  Currently
    # Module.params assumes that the initializer takes in a shape argument;
    # None acts as a flag to avoid shape checking.
    params = self.param('weightnorm_params', None, initializer)
    direction, scale, bias = [params[k] for k in ('direction', 'scale', 'bias')]
    return conv(inputs, _make_kernel(direction, scale)) + bias
```

This situation isn't terrible, but it does highlight the fact that the assumption that initializers depend on parameter shape and nothing else is a bit arbitrary.

A more flexible API, with `initializer` requiring _only_ a JAX PRNG key, would mean more consistent implementations of different types of initializers, and might also help to correctly emphasize what Flax's role is here, namely to handle splitting and passing of PRNG keys to initializers and to setup the parameters data-structure (a nested dictionary). 

### Proposal
We propose to change the call signature of `Module.param` from
```python
param(self, name: str, shape: Shape, initializer: Callable[[Key, Shape], Array]):
```
to
```python
param(self, name: str, initializer: Callable[[Key], Array]):
```
This change would lead to a slight simplification of the weightnorm example above, since the dummy shape arguments could be removed.

For existing Modules for which the initializer is a straightforward function of the parameter shape and no other data is required, we can alter the currying of the initializer definitions so that lines like
```python
kernel = self.param('kernel', kernel_shape, kernel_init)
```
can be replaced by
```python
kernel = self.param('kernel', kernel_init(kernel_shape))
```

There may be downsides to this approach which I, being a relative Flax noob, am unaware of. One thing is that we'd lose the shape checking in `Module.params`, but that seems like the kind of check which should be part of a test anyway.

### Alternatives
I think the obvious alternative is to simply keep the current API. The change proposed above is relatively minor but it still would likely require a number of users to make changes to their own code.

**Comment by jheek** [comment]
I agree the shape argument is arbitrary. It also implies that we can only return a single tensor which is also arbitrary.

Note that Flax is using the initializers from `jax.nn.initializers` so that create some difficulty when we want to reuse those.

An alternative that is less invasive is the following:

Change the current param api:
```
def param(self, shape, init_fn):
  ...
  init_value = init_fn(rng, shape)
```
to:
```
def param(self[, shape=None], init_fn):
  if shape is None:
    init_value = init_fn(rng):
  else:
    init_value = init_fn(rng, shape)
```

**Comment by j-towns** [comment]
Isn't 
```python
def param(self, shape=None, init_fn):
```
a SyntaxError?  Just wondering if there's a way around that...

**Comment by j-towns** [comment]
> Note that Flax is using the initializers from jax.nn.initializers so that create some difficulty when we want to reuse those.

I think it's pretty straightforward to wrap the JAX initializers though.

**Comment by jheek** [comment]
> Isn't
> 
> ```python
> def param(self, shape=None, init_fn):
> ```
> 
> a SyntaxError? Just wondering if there's a way around that...

Yes you have to implement some manual argument unpacking to get that working. With the brackets I meant the param method is basically overloaded so you have either

`param(name, shape, init_fn)`
or
`param(name, init_fn)`

we could also decide to give the second option a different name.

**Comment by avital** [comment]
My default inclination is to avoid non-stardard ways of defining optional
arguments, unless we explicitly say that one form is deprecated and only
there for backwards compatibility.

**Comment by jekbradbury** [comment]
If both Flax and Haiku want a different `jax.nn.initializers` API we should maybe switch it...

**Comment by levskaya** [comment]
@jekbradbury - oh, do haiku peeps also want a different initializer api?

**Comment by avital** [comment]
This has been resolved (with an improved shape check) in #530
